### DISCUSSION TITLE
sum-of-squares and semi-definite programming

### DISCUSSION THREAD

**Discussion by dio4ev** [yalmip/YALMIP#1018] [General discussion] [answered]
Dear Mr. Loefberg!

I am working on a static program analysis tool.
This tool gets some program (in a special syntax) and reduces the problem of finding a suitable runtime bound to checking positivity of polynomials.

First of all, Yalmip and its SOS methods are great for this task.
But, Matlab is apparently really slow, also within a compiled stand-alone application and slows down the analysis of my tool drastically.
As Yalmip is a Matlab application, I suppose that there is no binary application (without any Matlab interaction) available?

Now my idea is to build my own *simple* backend which transforms a polynomial into a suitable input for a given SDP solver and thus get faster results.
But I do not completely understand which transformations I need in order to do that and how, so with which input, I have to call the SDP solver.

Do you know some good literature on that?
Especially where it is explained how I have to write the SDP solver input (in SDPA format?) for a given polynomial and solver like CSDP?

Can I see on which semi-definite program Yalmip calls the SDP solver?

Until now I have read [1], glanced through [2] [3] and I plan to look at [4].

Thank you very much for your help!

[1]: https://link.springer.com/chapter/10.1007/978-3-540-74591-4_9
[2]: https://link.springer.com/article/10.1007/s10107-003-0387-5
[3]: https://www.stat.berkeley.edu/~jsteinhardt/stat260/notes/lect10.pdf
[4]: https://link.springer.com/chapter/10.1007/978-3-540-30579-8_1

**Comment by johanlofberg** [comment, marked as answer]
Nothings pops in my mind where the super low-level stuff is discussed. It's basically about converting the model in "Doing it your self" in https://yalmip.github.io/tutorial/sumofsquaresprogramming/ to a solver specific format (either in primal or dual form)

Reverse engineering with specific solver is probably what you start with

```
sdpvar x y t
p = x^4 + x*y + y^4
ops = sdpsettings('sos.newton',0,'sos.congruence',0,'sos.scale',0)
ops = sdpsettings(ops,'sos.model','image')
[F,h] = compilesos(sos(p-t),-t,ops)
F
sdisplay(sdpvar(F(1)))
M = export(F,h,sdpsettings('solver','sedumi'))

ops = sdpsettings(ops,'sos.model','kernel')
[F,h] = compilesos(sos(p-t),-t,ops)
M = export(F,h,sdpsettings('solver','sedumi'))


```

**Reply by dio4ev** [reply]
Thank you! I will have a look at it :)

**Comment by dio4ev** [comment]
Dear Mr Loefberg!

Sorry to necrobump an already answered discussion, but my question is a follow-up question to the one above.
 
I implemented a lot of stuff mostly following the Yalmip website, [0] and [1].
Unfortunately, I still can't solve the example constraint with my own implementation.

For example, my implementation reduces a specific problem to checking that the following polynomial is postive:
`- x+ 2 * (x*y)+ 3/2 * (x*a)- x*a*z- x^2+ 1/2 * (x^2*a)+ y- 3/2 * (y*a)- y*a*z- y^2+ 1/2 * (y^2*a)+ a*z^2`

Previously. with Yalmip and Matlab my tool generates the following matlab file on which matlab is then called and solves it:

```
yalmip('clear');

x = sdpvar(1,1);
i4650321219155950785 = sdpvar(1,1);
y = sdpvar(1,1);
z = sdpvar(1,1);

a = sdpvar(1,1);

ctxt = [- 1+ x- i4650321219155950785;- 1+ x- y;- 1+ i4650321219155950785- y;- y+ z;x- z;y];
p4650321219155950785 = - x+ 2 * (x*y)+ 3/2 * (x*a)- x*a*z- x^2+ 1/2 * (x^2*a)+ y- 3/2 * (y*a)- y*a*z- y^2+ 1/2 * (y^2*a)+ a*z^2;

[s0, c0] = polynomial([ x i4650321219155950785 y z],1);
[s1, c1] = polynomial([ x i4650321219155950785 y z],1);
[s2, c2] = polynomial([ x i4650321219155950785 y z],1);
[s3, c3] = polynomial([ x i4650321219155950785 y z],1);
[s4, c4] = polynomial([ x i4650321219155950785 y z],1);
[s5, c5] = polynomial([ x i4650321219155950785 y z],1);

F = [sos(p4650321219155950785 - [s0 s1 s2 s3 s4 s5] * ctxt),sos(s0),sos(s1),sos(s2),sos(s3),sos(s4),sos(s5)];
solvesos(F,[a],[],[c0;c1;c2;c3;c4;c5;a]);

display(a);
```
As I wrote in my previous question, I would like to get rid of the matlab dependency as the startup is incredible slow.

Currently I encode the SOS problem into an SDPA file [2] in a similar way (without optimizations like the ones in Yalmip) as in the Matlab code.
The problem is that CSDP binary [3] still returnes "infeasible" and I don't know if some optimizations are missing or I just encode the problem in an incorrect way.
My SDPA problem consists of 2 blocks per constraint matrix, where the first one chooses the correct variables for a monomial and the second one encodes the decision variables (=variables for which CSDP should find an optimal value such that the polynomial is SOS).
In the second matrix a decision variable variable is presented by two variables restricted to non-negative values, e.g. for the decision variable `a` in the polyniomial above we take `a = a1 - a2` and then encode it as follows for each monomial:
```
W=[Y    0   0;
      0   a1   0;
      0    0   a2 ]
```
where `W` is a constraint matrix [1] and `Y` is an `n x n matrix` and `n` is the number of possible monomials in the SOS representation.
For the monomial `x*a*z` with possible monomials `[1 x y z]` the matrix `W` would look as follows:
```
W=[0   1   1   0   0   0;
      1   0   1   0   0   0;
      1   1   0   0   0   0;
      0   0   0   0   0   0;
      0   0   0   0   -1   0;
      0   0   0   0   0   1]
```
I also investigated some techniques like, Newton polytope reduction, but this does not help in my case here.

Do you have an idea what I am doing wrong here or what I am missing?
Thank you very much for reading all this and your help :)

Kind regards,
Jonas

[0]: https://www.ime.usp.br/~fmario/sdp/sdpa.pdf
[1]: http://plato.asu.edu/ftp/sdpa_format.txt
[2]: https://pastebin.com/Jn9jh79j
[3]: https://github.com/coin-or/Csdp

**Reply by johanlofberg** [reply]
That's too low-level for me to see. I can only advice you to turn off all tricks such as congruence, newton, scale, and maybe switch to image model, in sdpsettings, and then export and try to reverse engineer to understand what comes out

**Reply by dio4ev** [reply]
Okay, thank you very much!

**Reply by dio4ev** [reply]
Which fields do I have to disable for that? [0]
From the documentation, it is not completely clear for me

[0]: https://yalmip.github.io/command/sdpsettings/

**Reply by johanlofberg** [reply]
they are all in the sos field, so just look at that to see what is available

**Reply by dio4ev** [reply]
Thank you! :)
